### ISSUE TITLE
Is Backpack really needed?

### ISSUE THREAD

**Issue by NicolasT** [ocharles/haskell-opentracing#2]
I was wondering whether the use of Backpack is really required. It may indeed provide an elegant way to split the API from implementations and allow for ad-hoc selection of implementations, though also comes with a cost: it's very new, tool support is not 100%, and requires a very recent version of GHC.

Did you consider an approach using type-classes (admittedly lawless ones) with some associated types? Here's what this could look like:

```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE TypeFamilies #-}

module Demo where

import Control.Lens
import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BS8
import Data.Map (Map)
import qualified Data.Map as Map
import Data.String (IsString)
import Data.Text (Text)
import qualified Data.Text as T

-- Library

newtype OperationName = OperationName { getOperationName :: Text }
    deriving (Show, Eq, IsString)

newtype Timestamp = Timestamp { getTimestamp :: Word }
    deriving (Show, Eq, Ord)

data Reference ctx = ChildOf ctx
                   | FollowsFrom ctx
    deriving (Show, Eq)

type TagName = Text

data TagValue = TagString Text
              | TagBool Bool
              | TagNumeric Int

data SpanState = Started | Finished

class Span (span :: SpanState -> *) spanContext | span -> spanContext where
    spanContext :: Getter (span r) spanContext
    -- The OpenTracing spec is very fuzzy when it comes to defining access. I guess having a
    -- operationName' :: Getter (span r) OperationName
    -- would make sense for read-only purposes (similar for tags, references,...)
    operationName :: Lens' (span 'Started) OperationName
    setTag :: TagName -> TagValue -> span 'Started -> span 'Started
    -- TODO Log structured data
    -- TODO Baggage stuff

data Format f where
    TextMap :: Format (Map ByteString ByteString)
    HTTPHeaders :: Format (Map Text Text)
    Binary :: Format ByteString

class Span (TracerSpan t) (TracerSpanContext t) => Tracer t where
    type TracerSpan t :: SpanState -> *
    type TracerSpanContext t

    startSpan :: t
              -> OperationName
              -> [Reference (TracerSpanContext t)]
              -> Maybe Timestamp
              -> Map TagName TagValue
              -> IO (TracerSpan t 'Started)

    finishSpan :: t
               -> TracerSpan t 'Started
               -> Maybe Timestamp
               -> IO (TracerSpan t 'Finished)

    inject :: t
           -> Format c
           -> c
           -> TracerSpanContext t
           -> c

    extract :: t
            -> Format c
            -> c
            -- I bet this should be 'Maybe (TracerSpanContext t)'
            -> TracerSpanContext t

    emitSpan :: t
             -> TracerSpan t 'Finished
             -> IO ()


-- Implementation
data Jaeger = Jaeger

data JaegerSpan' = JaegerSpan' { jaegerSpanSpanContext :: JaegerSpanContext
                               , jaegerSpanOperationName :: OperationName
                               , jaegerSpanTags :: Map TagName TagValue
                               , jaegerSpanReferences :: [Reference JaegerSpanContext]
                               , jaegerSpanStartTime :: Timestamp
                               }

data JaegerSpan (s :: SpanState) where
    JaegerSpanActive :: JaegerSpan' -> JaegerSpan 'Started
    JaegerSpanFinished :: JaegerSpan' -> Timestamp -> JaegerSpan 'Finished

data JaegerSpanContext = JaegerSpanContext

mkJaegerSpanContext :: IO JaegerSpanContext
mkJaegerSpanContext = return JaegerSpanContext

getTime :: IO Timestamp
getTime = return (Timestamp 0)

instance Span JaegerSpan JaegerSpanContext where
    spanContext = to $ \case
        JaegerSpanActive s -> jaegerSpanSpanContext s
        JaegerSpanFinished s _ -> jaegerSpanSpanContext s
    operationName = lens
        (\(JaegerSpanActive s) -> jaegerSpanOperationName s)
        (\(JaegerSpanActive s) n -> JaegerSpanActive s { jaegerSpanOperationName = n })
    setTag name value (JaegerSpanActive s) = JaegerSpanActive s { jaegerSpanTags = Map.insert name value (jaegerSpanTags s) }

instance Tracer Jaeger where
    type TracerSpan Jaeger = JaegerSpan
    type TracerSpanContext Jaeger = JaegerSpanContext

    startSpan Jaeger o r ts t = do
        ctx <- mkJaegerSpanContext
        start <- maybe getTime return ts
        return $ JaegerSpanActive JaegerSpan' { jaegerSpanSpanContext = ctx
                                              , jaegerSpanOperationName = o
                                              , jaegerSpanTags = t
                                              , jaegerSpanReferences = r
                                              , jaegerSpanStartTime = start
                                              }

    finishSpan Jaeger (JaegerSpanActive s) ts = do
        end <- maybe getTime return ts
        return $ JaegerSpanFinished s end

    emitSpan Jaeger _ = error "Not implemented"

    -- TODO Actual serialization and parsing
    inject Jaeger format carrier _ctx = case format of
        TextMap -> Map.insert (BS8.pack "jaegerContext") BS.empty carrier
        HTTPHeaders -> Map.insert (T.pack "X-Jaeger-Context") T.empty carrier
        Binary -> BS.append carrier BS.empty

    extract Jaeger format _carrier = case format of
        TextMap -> JaegerSpanContext
        HTTPHeaders -> JaegerSpanContext
        Binary -> JaegerSpanContext
```

Not saying this is the way to go, if you believe Backpack is the way to go then so be it :smile: Just starting a little discussion.

**Comment by ocharles** [comment]
This is a fairly philosophical question, which has the immediate answer of - no. Of course the full answer is no, because we managed to get this far in Haskell without Backpack!

However, I think the presentation above is significantly more complicated than the Backpack solution, and presenting the implementation is only half the puzzle. The rest is making sure that it's both a) fairly painless to consume and b) painless to type check. The latter can become very difficult when you've got all sorts of associated types flying around.

In the Backpack world, I was amazed at just how simple the client facing API is. It just looks like plain old Haskell! If you really need features that only one implementation has, then your library depends on that implementation, and not the `opentracing-client` library. Others don't have to make the same commitment, later on when you fill the holes in, it all "just works".

Backpack also has the rather nice advantage of actually inlining well. There are no dictionaries, it's just normal Haskell - and GHC does a darn good job at optimising that.

So in short, no, Backpack is certainly not required, but this project was started as a playground for me to learn Backpack (as the same as Kim), but I feel like it's a pretty good fit for it.

To turn the question around, what does your non-Backpack solution do (other than working on < 8.2) that Backpack can't?

**Comment by NicolasT** [comment]
Fair enough :+1: Actually, the only 'benefit' I saw was support for GHC < 8.2 :wink:

I guess that, as long as only `opentracing-client` 'requires' Backpack, and `jaeger-client` and others remain 'traditional' Haskell packages, those who want e.g. Jaeger support but can't depend on GHC >= 8.2 (yet) can still get things to work, or one could implement a generalized lib as I sketched above, outside of the realm of this specific project.

**Comment by ocharles** [comment]
Right, that's a great point, too!

**Comment by ocharles** [comment]
I'm going to rudely at @ezyang, just in case he thinks I missed any of his points. BTW, it's worth reading the thesis for motivation too, but I'm drinking up as much of the Kool Aid as possible.

**Comment by ezyang** [comment]
I think @ocharles has hit the nail on the head, including the fact that implementation modules tend not to be Backpack specific, and can thus be used independently.

The big thing that you lose if you are not on GHC 8.2 is the ability to make use of any modules that are parameterized on the implementation: `opentracing-client`, in this case. In principle, it should be possible to do some source code generation to make a GHC 8.0 (and earlier) version of the library... but no one has written the code yet :0
